<div class="alert alert-block alert-success">
<b>Comentario general del revisor</b> <a class="tocSkip"></a><br />
Status del proyecto: <b>Aprobado</b>
</div>


¡Hola!<br />
Soy **Francisco Cortés**, estoy contento de revisar tu proyecto y ser parte de tu proceso de aprendizaje.
A lo largo del texto, haré algunas observaciones sobre mejoras en el código y también haré comentarios sobre tus percepciones sobre el tema. Si existe algún error en el código, no te preocupes, estoy aquí para ayudarte a mejorarlo, en la primera iteración te lo señalaré para que tengas la oportunidad de corregirlo, pero si aún no encuentras una solución para esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a retroalimentación y discusiones sobre el tema.<br />
Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o borres**.
Revisaré cuidadosamente tu código para comprobar que se han cumplido con los requisitos y te proporcionaré mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si la ejecución fue perfecta succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si existe alguna recomendación para que tu código mejore.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si existen correcciones necesarias para cumplir con los requisitos. El trabajo no puede ser aceptado si hay alguna caja roja.
</div>

Puedes responderme de la siguiente manera:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante.</b> <a class="tocSkip"></a>
</div>

# Descripción del proyecto

Trabajas en la compañía de extracción de petróleo OilyGiant. Tu tarea es encontrar los mejores lugares donde abrir **200 pozos nuevos de petróleo.**

Para completar esta tarea, tendrás que realizar los siguientes pasos:

- Leer los archivos con los parámetros recogidos de pozos petrolíferos en la región seleccionada: calidad de crudo y volumen de reservas.
- Crear un modelo para predecir el volumen de reservas en pozos nuevos.
- Elegir los pozos petrolíferos que tienen los valores estimados más altos.
- Elegir la región con el beneficio total más alto para los pozos petrolíferos seleccionados.

Tienes datos sobre muestras de crudo de tres regiones. Ya se conocen los parámetros de cada pozo petrolero de la región. Crea un modelo que ayude a elegir la región con el mayor margen de beneficio. Analiza los beneficios y riesgos potenciales utilizando la técnica *bootstrapping*.

## Condiciones

- Solo se debe usar la regresión lineal para el entrenamiento del modelo.
- Al explorar la región, se lleva a cabo un estudio de 500 puntos con la selección de los mejores 200 puntos para el cálculo del beneficio.
- El presupuesto para el desarrollo de 200 pozos petroleros es de 100 millones de dólares.
- Un barril de materias primas genera 4.5 USD de ingresos. El ingreso de una unidad de producto es de 4500 dólares (el volumen de reservas está expresado en miles de barriles).
- Después de la evaluación de riesgo, mantén solo las regiones con riesgo de pérdidas inferior al 2.5%. De las que se ajustan a los criterios, se debe seleccionar la región con el beneficio promedio más alto.

Los datos son sintéticos: los detalles del contrato y las características del pozo no se publican.

## Descripción de datos

Los datos de exploración geológica de las tres regiones se almacenan en archivos:

- */datasets/geo_data_0.csv*. Descarga el conjunto de datos https://practicum-content.s3.us-west-1.amazonaws.com/datasets/geo_data_0.csv
- */datasets/geo_data_1.csv*. Descarga el conjunto de datos https://practicum-content.s3.us-west-1.amazonaws.com/datasets/geo_data_1.csv
- */datasets/geo_data_2.csv*. Descarga el conjunto de datos https://practicum-content.s3.us-west-1.amazonaws.com/datasets/geo_data_2.csv 
- *id* — identificador único de pozo de petróleo
- *f0, f1, f2* — tres características de los puntos (su significado específico no es importante, pero las características en sí son significativas)
- *product* — volumen de reservas en el pozo de petróleo (miles de barriles).

## Instrucciones del proyecto
1. Descarga y prepara los datos. Explica el procedimiento.
2. Entrena y prueba el modelo para cada región en geo_data_0.csv:
    1. Divide los datos en un conjunto de entrenamiento y un conjunto de validación en una proporción de 75:25
    2. Entrena el modelo y haz predicciones para el conjunto de validación.
    3. Guarda las predicciones y las respuestas correctas para el conjunto de validación.
    4. Muestra el volumen medio de reservas predicho y RMSE del modelo.
    5. Analiza los resultados.
    6. Coloca todos los pasos previos en funciones, realiza y ejecuta los pasos 2.1-2.5 para los archivos 'geo_data_1.csv' y 'geo_data_2.csv'.
3. Prepárate para el cálculo de ganancias:
    1. Almacena todos los valores necesarios para los cálculos en variables separadas.
    2. Dada la inversión de 100 millones por 200 pozos petrolíferos, de media un pozo petrolífero debe producir al menos un valor de 500,000 dólares en unidades para evitar pérdidas (esto es equivalente a 111.1 unidades). Compara esta cantidad con la cantidad media de reservas en cada región.
    3. Presenta conclusiones sobre cómo preparar el paso para calcular el beneficio.
4. Escribe una función para calcular la ganancia de un conjunto de pozos de petróleo seleccionados y modela las predicciones:
    1. Elige los 200 pozos con los valores de predicción más altos de cada una de las 3 regiones (es decir, archivos 'csv').
    2. Resume el volumen objetivo de reservas según dichas predicciones. Almacena las predicciones para los 200 pozos para cada una de las 3 regiones.
    3. Calcula la ganancia potencial de los 200 pozos principales por región. Presenta tus conclusiones: propón una región para el desarrollo de pozos petrolíferos y justifica tu elección.
5. Calcula riesgos y ganancias para cada región:
    1. Utilizando las predicciones que almacenaste en el paso 4.2, emplea la técnica del bootstrapping con 1000 muestras para hallar la distribución de los beneficios.
    2. Encuentra el beneficio promedio, el intervalo de confianza del 95% y el riesgo de pérdidas. La pérdida es una ganancia negativa, calcúlala como una probabilidad y luego exprésala como un porcentaje.
    3. Presenta tus conclusiones: propón una región para el desarrollo de pozos petrolíferos y justifica tu elección. ¿Coincide tu elección con la elección anterior en el punto 4.3?

## Evaluación del proyecto
Hemos definido los criterios de evaluación para el proyecto. Lee esto con atención antes de pasar al ejercicio.

Esto es lo que los revisores buscarán cuando evalúen tu proyecto:

- ¿Cómo preparaste los datos para el entrenamiento?
- ¿Seguiste todos los pasos de las instrucciones?
- ¿Consideraste todas las condiciones del negocio?
- ¿Cuáles son tus hallazgos sobre el estudio de tareas?
- ¿Aplicaste correctamente la técnica bootstrapping?
- ¿Sugeriste la mejor región para el desarrollo de pozos? ¿Justificaste tu elección?
- ¿Evitaste la duplicación de código?
- ¿Mantuviste la estructura del proyecto y el código limpio?

Ya tienes tus hojas informativas y los resúmenes de los capítulos, por lo que todo está listo para continuar con el proyecto.

# Proyecto

Como siempre, iniciemos importando mis librerias.

In [13]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import root_mean_squared_error
from sklearn.preprocessing import StandardScaler

In [ ]:
# Importo mis datasets como df_0, df_1 y df_2
df_0 = pd.read_csv('datasets/geo_data_0.csv')
df_1 = pd.read_csv('datasets/geo_data_1.csv')
df_2 = pd.read_csv('datasets/geo_data_2.csv')

In [5]:
df_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

## Análisis Exploratorio

In [6]:
pd.set_option('display.float_format', '{:,.2f}'.format)

display(
    df_0.info(),
    df_0.describe(),
    print(f'\nEl dataframe 0 tiene una cantidad de {df_0.duplicated().sum()} valores duplicados')
)
display(
    df_1.info(),
    df_1.describe(),
    print(f'\nEl dataframe 1 tiene una cantidad de {df_0.duplicated().sum()} valores duplicados')
)
display(
    df_2.info(),
    df_2.describe(),
    print(f'\nEl dataframe 2 tiene una cantidad de {df_0.duplicated().sum()} valores duplicados')
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB

El dataframe 0 tiene una cantidad de 0 valores duplicados


None

f0         f1         f2    product
count 100,000.00 100,000.00 100,000.00 100,000.00
mean        0.50       0.25       2.50      92.50
std         0.87       0.50       3.25      44.29
min        -1.41      -0.85     -12.09       0.00
25%        -0.07      -0.20       0.29      56.50
50%         0.50       0.25       2.52      91.85
75%         1.07       0.70       4.72     128.56
max         2.36       1.34      16.00     185.36

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB

El dataframe 1 tiene una cantidad de 0 valores duplicados


None

f0         f1         f2    product
count 100,000.00 100,000.00 100,000.00 100,000.00
mean        1.14      -4.80       2.49      68.83
std         8.97       5.12       1.70      45.94
min       -31.61     -26.36      -0.02       0.00
25%        -6.30      -8.27       1.00      26.95
50%         1.15      -4.81       2.01      57.09
75%         8.62      -1.33       4.00     107.81
max        29.42      18.73       5.02     137.95

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB

El dataframe 2 tiene una cantidad de 0 valores duplicados


None

f0         f1         f2    product
count 100,000.00 100,000.00 100,000.00 100,000.00
mean        0.00      -0.00       2.50      95.00
std         1.73       1.73       3.47      44.75
min        -8.76      -7.08     -11.97       0.00
25%        -1.16      -1.17       0.13      59.45
50%         0.01      -0.01       2.48      94.93
75%         1.16       1.16       4.86     130.60
max         7.24       7.84      16.74     190.03

None

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>

Buena manera de leer los datos y de hacer una revisión inicial
</div>


Los tres datasets cuentan con 100k rows de datos, no hay datos ausentes por lo tanto no aplicaremos metodos de imputación. Tampoco hay valores duplicados. Estamos listos para los siguientes pasos.

Para entrenar el modelo, sera necesario hacer drop a mi columna de id, ya que no me aporta nada en particular y necesitaria de NLP para poderla procesar.

## Función para Modelo de Regresion Lineal

In [ ]:
def region_model(df):
    '''
    Esta funcion recibe como entrada un dataframe.

    Una ves recibiendo esto se le da el procesamiento completo de dropear las columnas, preparas los datos en test y train, se entrena un modelo de LinearRegression 
    '''
    X = df.drop(columns=['id', 'product'])
    y = df['product']

    # Dividimos los datos en conjunto de entrenamiento (75%) y prueba (25%)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.25, random_state=42
    )   

    # Creamos un Pipeline con StandardScaler y Regresion Lineal (Usar Ridge o Lasso) 
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('regressor', LinearRegression())
    ])

    # Entrenamos el modelo
    pipeline.fit(X_train, y_train)
    
    # Hacemos predicciones
    y_pred = pipeline.predict(X_test)

    #Calculamos las métricas
    rmse = root_mean_squared_error(y_test, y_pred)
    relative_error = rmse / y_test.mean()

    # Mostramos valor del RMSE, Error Relativo y Valor Medio de las reservas predicted
    print(f"🔹 Root Mean Squared Error (RMSE): {rmse:.2f}")
    print(f"📏 Error relativo: {relative_error:.2%}")
    print(f"🛢️ Volumen medio de reservas: {y_pred.mean():.2f} barriles")

    return 

<div class="alert alert-block alert-success">
<b>Comentario delxrevisor</b> <a class="tocSkip"></a><br />
Bien hecho!<br/>

Buena manera de construir una función para entrenar y validar el modelo
</div>


In [79]:
df_list = [df_0, df_1, df_2]
number = 0

for i in df_list:
    print(f'DataFrame {number}')
    region_model(i)
    print('\n')
    number += 1

DataFrame 0
🔹 Root Mean Squared Error (RMSE): 37.76
📏 Error relativo: 40.89%
🛢️ Volumen medio de reservas: 92.40 barriles


DataFrame 1
🔹 Root Mean Squared Error (RMSE): 0.89
📏 Error relativo: 1.30%
🛢️ Volumen medio de reservas: 68.71 barriles


DataFrame 2
🔹 Root Mean Squared Error (RMSE): 40.15
📏 Error relativo: 42.19%
🛢️ Volumen medio de reservas: 94.77 barriles




## Cálculo de Ganancias

In [75]:
# Valores
pozos = 200
budget = 100000000
precio_barril = 4500
minimo_promedio = int(budget/pozos)
minimo_barril = minimo_promedio/precio_barril

print(f'💰 El valor minimo que debe generar cada pozo en promedio es de ${minimo_promedio} USD')
print(f'🛢️ Cada pozo debe producir un minimo de {int(minimo_barril)} barriles')

💰 El valor minimo que debe generar cada pozo en promedio es de $500000 USD
🛢️ Cada pozo debe producir un minimo de 111 barriles


### Media de barriles por region
Region 1: 92.40  
Region 2: 68.71  
Region 3: 94.77

Ni una de las 3 regiones genera un promedio minimo de 111 barriles, por lo que si elegimos 200 pozos aleatoriamente ninguna región sera rentable.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Correcto!<br/>

Me parece acertada la manera de mostrar los resultados para cada una de las regiones, estoy de acuerdo con las observaciones dadas, ninguna de las regiones es rentable si se toman de esta manera
</div>


## Ganancia Potencial
El modelo pronostica la cantidad de barriles de petroleo (en miles) por cada pozo. De todos los pronosticos realizados debo elegir los 200 pozos mas grandes. Modifico mi funcion y regreso las 200 predicciones mas grandes.

In [104]:
# Modifico mi funcion region_model para extraer el top 200 de cada pozo. A su vez, modifico mi y_pred para que tome en cuenta todo el dataset y no solamente X_test
def region_model(df):
    '''
    Esta funcion recibe como entrada un dataframe.

    Una ves recibiendo esto se le da el procesamiento completo de dropear las columnas, preparas los datos en test y train, se entrena un modelo de LinearRegression 
    '''
    X = df.drop(columns=['id', 'product'])
    y = df['product']

    # Dividimos los datos en conjunto de entrenamiento (75%) y prueba (25%)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.25, random_state=42
    )   

    # Creamos un Pipeline con StandardScaler y Regresion Lineal (Usar Ridge o Lasso) 
    pipeline = Pipeline([
        ('scaler', StandardScaler()),
        ('regressor', LinearRegression())
    ])

    # Entrenamos el modelo
    pipeline.fit(X_train, y_train)
    
    # Hacemos predicciones SOBRE mi dataset completo
    y_pred = pipeline.predict(X)

    # Extraigo el top 200 pozos de mis predicciones basado en producto
    top_200 = pd.Series(y_pred).sort_values(ascending=False).head(200)

    return top_200

In [105]:
# Corro mi funcion con cada DF para extraer mi top 200 pozos por region en formato de Pandas Series
region_0_top_200 = region_model(df_0)
region_1_top_200 = region_model(df_1)
region_2_top_200 = region_model(df_2)

In [116]:
# Ganancia Potencial

print(f'🛢️ La region 1 tiene un promedio de barriles de: {region_0_top_200.mean()} y una ganancia potencial de ${region_0_top_200.sum()*precio_barril:.2f} USD\n')
print(f'🛢️ La region 2 tiene un promedio de barriles de: {region_1_top_200.mean()} y una ganancia potencial de ${region_1_top_200.sum()*precio_barril:.2f} USD\n')
print(f'🛢️ La region 3 tiene un promedio de barriles de: {region_2_top_200.mean()} y una ganancia potencial de ${region_2_top_200.sum()*precio_barril:.2f} USD\n')

🛢️ La region 1 tiene un promedio de barriles de: 163.34728730628322 y una ganancia potencial de $147012558.58 USD

🛢️ La region 2 tiene un promedio de barriles de: 139.16109357667088 y una ganancia potencial de $125244984.22 USD

🛢️ La region 3 tiene un promedio de barriles de: 156.59273115041017 y una ganancia potencial de $140933458.04 USD



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Bien hecho!<br/>

Buena manera de encontrar los resultados
</div>


### Conclusion

Basandome en las predicciones del modelo para cada una de las regiones. La región que propongo para el desarrollo de los nuevos pozos petroleros es la región 1, ya que tiene el pronóstico de promedio de barriles mas alto. Si bien las 3 cumplen con el quota, la region 1 es la que tiene potencialmente mayores y mejores ganancias.

5. Calcula riesgos y ganancias para cada región:
    1. Utilizando las predicciones que almacenaste en el paso 4.2, emplea la técnica del bootstrapping con 1000 muestras para hallar la distribución de los beneficios.
    2. Encuentra el beneficio promedio, el intervalo de confianza del 95% y el riesgo de pérdidas. La pérdida es una ganancia negativa, calcúlala como una probabilidad y luego exprésala como un porcentaje.
    3. Presenta tus conclusiones: propón una región para el desarrollo de pozos petrolíferos y justifica tu elección. ¿Coincide tu elección con la elección anterior en el punto 4.3?

In [ ]:
def bootstrap(df, n_bootstraps = 1000):
    ''' 
    Realiza un bootstrapping sobre un dataframe completo y estima
    la distribucion de ganancias, intervalo de confianza y riesgo de pérdida.
    '''

    # Hago mis variables X y y
    X = df.drop(columns=['id', 'product'])
    y = df['product']

    # Necesito agregar un random state para mi for loop
    state = np.random.RandomState(42)
    profit = []
    
    # Inicio un for loop para mi bootstrap, lo aplico para ganancias obtenidas para los 200 pozos
    for i in range(n_bootstraps):
        sample = y.sample(n=200, replace = True, random_state=state)
        total_sample = pd.Series(sample).sum()
        ingreso = total_sample * precio_barril
        ganancia = ingreso - budget
        profit.append(ganancia)
    
    # Lo paso a serie de Pandas
    profit = pd.Series(profit)
    risk = (profit < 0).mean()

    lower = profit.quantile(0.05)
    upper = profit.quantile(0.95)

    print(f'💰 Ganancia Promedio: ${profit.mean():.2f}')
    print(f'📈 Intervalo de confianza (95%): Entre ${lower:.2f} a ${upper:.2f}')
    print(f'🛑 El riesgo de perdida es del {risk*100:.2f}%')
    
    return 

Calculo el riesgo de perdida como un porcentaje de cuantas ocasiones mi bootstrap genero menos de 0, es decir mi ingreso - budget. Budget siendo 100m

In [ ]:
print('Region 1')
bootstrap(df_0)
print('\nRegion 2')
bootstrap(df_1)
print('\nRegion 3')
bootstrap(df_2)

Region 1
💰 Ganancia Promedio: $47018969.53
📈 Intervalo de confianza (95%): Entre $46430800.51 a $47662940.71
🛑 El riesgo de perdida es del 0.00%

Region 2
💰 Ganancia Promedio: $25245181.58
📈 Intervalo de confianza (95%): Entre $25219343.08 a $25272580.26
🛑 El riesgo de perdida es del 0.00%

Region 3
💰 Ganancia Promedio: $40938345.13
📈 Intervalo de confianza (95%): Entre $40397017.48 a $41514500.66
🛑 El riesgo de perdida es del 0.00%


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Cuidado!<br/>

Estas mandando los datos de los mejores 200 pozos de cada región y la función de bootstrap lo que hace es tomar una muestra aleatoria de 200 pozos, por lo que todo el tiempo esta tomando los mismos 200 pozos para hacer el calculo y resulta en buenas ganancias con un riesgo de 0, lo que se debe hacer es mandar los datos completos para que se tomen 200 pozos aleatorios de todos los existentes
</div>


## Conclusión

Basado en el análisis de rentabilidad y riesgo, se recomienda invertir en la **Region 1**:
- Presenta mayor beneficio (ganancia) promedio.
- Riesgo de perdida de 0%.
- Intervalo de confianza con mayor rendimiento.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br />
Has realizado un buen trabajo, me doy cuenta de que has aplicado los conocimientos que has adquirido durante el curso y los procedimientos realizados son correctos, esta es una gran herramienta que nos ayuda en muchos casos y es bastante útil en proyectos reales.<br/>
    
El documento que entregas es organizado y fácil de entender.<br/>

Este tipo de analisis nos ayuda a confirmar suposiciones o apoyar decisiones de negocio las cuales muchas veces entre más puntos de vista se analisen mejores resultados se obtendrán al final.
Continúa con el buen trabajo y mucho éxito en el siguiente Sprint!
</div>
